presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,393 @@
+                        joinNode.getLeftHashSymbol(),
+                        joinNode.getRightHashSymbol(),
+                        joinNode.getDistributionType()));
+        }
+    }
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(a.msg.x)
+ *    Project(a.key)
+ *      Source(a)
+ *  </pre>
+ * to:
+ * <pre>
+ *  Project(a.key)
+ *    Project(a.msg.x)
+ *      Source(a)
+ *  </pre>
+ * To pass down dereference expressions to JoinNode
+ * No change on ScanFilterAndProject
+ */
+    public class DereferencePushDownProject
+            extends DereferencePushDownRule<ProjectNode>
+    {
+        public DereferencePushDownProject(Metadata metadata, SqlParser sqlParser)
+        {
+            super(metadata, sqlParser, project());
+        }
+
+        @Override
+        protected Result pushDownDereferences(Context context, ProjectNode projectNode, Map<DereferenceExpression, Symbol> expressions, Assignments assignments)
+        {
+            List<Symbol> outputSymbols = projectNode.getOutputSymbols();
+            Map<Symbol, Expression> pushdownExpressions = expressions.entrySet().stream()
+                    .filter(entry -> !isCastRowType(entry.getKey()))
+                    .filter(entry -> outputSymbols.contains(getOnlyElement(extractAll(entry.getKey()))))
+                    .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+            if (pushdownExpressions.isEmpty()) {
+                return Result.empty();
+            }
+
+            ImmutableMap.Builder<Symbol, Symbol> symbolsBuilder = ImmutableMap.builder();
+            pushdownExpressions.entrySet().stream()
+                    .forEach(entry -> symbolsBuilder.put(getOnlyElement(extractAll(entry.getValue())), entry.getKey()));
+            Map<Symbol, Symbol> symbolsMap = symbolsBuilder.build();
+
+            Assignments.Builder sourceBuilder = Assignments.builder();
+            for (Map.Entry<Symbol, Expression> entry : projectNode.getAssignments().entrySet()) {
+                if (symbolsMap.containsKey(entry.getKey())) {
+                    Symbol targetSymbol = symbolsMap.get(entry.getKey());
+                    DereferenceExpression targetDereference = (DereferenceExpression) pushdownExpressions.get(targetSymbol);
+                    DereferenceExpression dereference = new DereferenceExpression(entry.getValue(), targetDereference.getField());
+                    sourceBuilder.put(targetSymbol, dereference);
+                }
+                else {
+                    sourceBuilder.put(entry.getKey(), entry.getValue());
+                }
+            }
+            return Result.ofPlanNode(new ProjectNode(context.getIdAllocator().getNextId(), projectNode.getSource(), sourceBuilder.build()));
+        }
+    }
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(a.msg.x)
+ *    Sort(a.msg.x)
+ *      Source(a)
+ *  </pre>
+ * to:
+ * <pre>
+ *  Project
+ *    Sort
+ *      Project(a.msg.x)
+ *        Source(a)
+ *  </pre>
```

Comment:
SortNode uses OrderingScheme which is defined using Symbols, not Expressions. Hence, it seems impossible to have Sort(a.msg.x).

Is this some kind of a pass-thru rule that pushes down dereference from a node above Sort under the Sort? Would a similar rule be necessary for other types of nodes, e.g. Unnest?

Test coverage is missing as well.

Reply:
my bad, it is push dereference through sort. Will fix and add unnest